### PR TITLE
Fix TransactAsset Implementation

### PR DESCRIPTION
**Pull request by apopiak**
This PR aims to fix the `TransactAsset` (tuple) implementation.
It converts local `AssetNotFound` errors into `XcmError::AssetNotFound` to allow the tuple implementation of TransactAsset to iterate properly. It also adjusts the other tuple implementations (notably `withdraw_asset` and `deposit_asset`) to have the same semantics.

Also adds some doc comments to XCM errors.

### CODE REVIEW

**Review by gavofyork — APPROVED** [review]

**Comment by gavofyork** [comment]
Would be good to add tests with this.

**Comment by apopiak** [comment]
As mentioned in chat, maybe we should remove the tuple implementation of `TransactAsset` to avoid the associated footguns?

**Comment by apopiak** [comment]
I adjusted the `TransactAsset` tuple implementation to something that I believe will be less error-prone.

**Comment by apopiak** [comment]
> Would be good to add tests with this.

Do you mean integration tests?
This only shows up for the tuple implementation.

File: xcm/xcm-executor/src/traits/transact_asset.rs [hunk 1]
```diff
@@ -130,33 +130,42 @@ impl TransactAsset for Tuple {
...
 		);
-		Err(XcmError::Unimplemented)
+		Err(XcmError::AssetNotFound)
 	}
 
 	fn transfer_asset(what: &MultiAsset, from: &MultiLocation, to: &MultiLocation) -> Result<Assets, XcmError> {
 		for_tuples!( #(
-			match Tuple::transfer_asset(what, from, to) { o @ Ok(_) => return o, _ => () }
+			match Tuple::transfer_asset(what, from, to) {
+				Err(XcmError::AssetNotFound) => (),
```
Comment (shawntabrizi) [inline xcm/xcm-executor/src/traits/transact_asset.rs:166]: Let's turn this into `Unimplemented` and `AssetNotFound`

And lets keep the errors above `Unimplemented`

**Comment by apopiak** [comment]
@shawntabrizi Another issue I see with the Tuple impl is that the `check_in` and `check_out` implementations have no way of signalling to each other, so an asset might be checked in twice which might or might not be what you want.
(Of course your point still stands that people can write their own combinator that does the correct thing.)

File: xcm/src/v0/traits.rs [hunk 2]
```diff
@@ -43,10 +46,15 @@ pub enum Error {
 	FailedToDecode,
 	BadOrigin,
 	ExceedsMaxMessageSize,
+	/// An asset transaction (like withdraw or deposit) failed.
+	/// See implementors of the `TransactAsset` trait for sources.
```
Comment (ordian) [inline xcm/src/v0/traits.rs:50]: spotted by the spellcheck CI check
```suggestion
	/// See implementers of the `TransactAsset` trait for sources.
```

**Review by shawntabrizi — APPROVED** [review]

**Comment by shawntabrizi** [comment]
I thought the "signaling" would be that if `check_in` happens successfully, it stops iteration?

**Comment by apopiak** [comment]
@gavofyork Fine to merge?

**Comment by apopiak** [comment]
bot merge

**Pull request merged**